tests/org.eclipse.ui.tests/Eclipse UI Tests/org/eclipse/ui/tests/keys/DispatcherTest.java

```diff
@@ -91,4 +97,44 @@ public void testDispatcherMultipleCommandsOnlyOneEnabled() throws Exception {
 		}
 	}
 
+	@Test
+	public void testSelectionVariableHandlerConflicts() throws Exception {
+		IFile file = p.getFile("test.whatever");
+		try (ByteArrayInputStream stream = new ByteArrayInputStream("hello".getBytes())) {
+			file.create(stream, true, new NullProgressMonitor());
+		}
+		CheckInvokedHandler.invoked = false;
+		IViewPart projectExplorer = PlatformUI.getWorkbench().getActiveWorkbenchWindow().getActivePage()
+				.showView(IPageLayout.ID_PROJECT_EXPLORER);
+
+		// select a file in the project explorer and assert that CTRL+C invokes our test
+		// handler
+		{
+			projectExplorer.getSite().getSelectionProvider().setSelection(new StructuredSelection(file));
+			dispatcher.press(Arrays.asList(KeyStroke.getInstance(SWT.CTRL, 'C')), null);
+			// in the Project Explorer, the custom copy handler should have been called
+			Assert.assertTrue("Handler should have been invoked", CheckInvokedHandler.invoked);
+		}
+
+		// invalidate the 'selection' variable and trigger copy operation again
+		{
+			// simulate that some user or platform code triggered a change in the evaluation
+			// context
+			// in real scenario, this would be opening a new Shell, for example
+			{
+				IEclipseContext ectx = projectExplorer.getSite().getService(IEclipseContext.class);
+				ectx.set(ISources.ACTIVE_CURRENT_SELECTION_NAME, null);
```

Comment:
And then used here?

Reply:
in the previous location, the "selection" variable was not filtered anymore... here I am resetting its value which is the whole point of this patch

Reply:
So `ISources.ACTIVE_CURRENT_SELECTION_NAME` is the actual initial problem, and changing the evaluationService is the solution?

Reply:
well, the evaluation context (containing the selection variable) is changed and the IEvaluationService is used which was not the case previously... so, there are two bugs from my point of view:
- use IEvaluationService to get a more accurate IEvaluationContext than just wrapping the IEclipseContext
- in the IEvaluationService listen consider changes of the selection variables which are triggered by legacy mechanisms